common/cpp/src/google_smart_card_common/value_conversion.h

```diff
@@ -27,21 +27,153 @@
+  EnumToValueConverter& operator=(const EnumToValueConverter&) = delete;
+  ~EnumToValueConverter();
+
+  void HandleItem(int64_t enum_item, const char* enum_item_name);
+  bool TakeConvertedValue(const char* type_name, Value* converted_value,
+                          std::string* error_message = nullptr);
+
+ private:
+  const int64_t enum_to_convert_;
+  optional<Value> converted_value_;
+};
+
+// Visitor of enum type's items that converts a string `Value` into a C++ enum
+// value, by finding the corresponding item among visited ones. This is a
+// non-template class, because enum values are casted to integers.
+class EnumFromValueConverter final {
+ public:
+  explicit EnumFromValueConverter(Value value_to_convert);
+  EnumFromValueConverter(const EnumFromValueConverter&) = delete;
+  EnumFromValueConverter& operator=(const EnumFromValueConverter&) = delete;
+  ~EnumFromValueConverter();
+
+  void HandleItem(int64_t enum_item, const char* enum_item_name);
+  bool GetConvertedEnum(const char* type_name, int64_t* converted_enum,
+                        std::string* error_message) const;
+
+ private:
+  const Value value_to_convert_;
+  optional<int64_t> converted_enum_;
+};
+
+}  // namespace internal
+
+///////////// EnumValueDescriptor ////////////
+
+// Class that allows to describe an enum's items and their corresponding string
+// names. Can be used in order to automatically implement conversion of a C++
+// enum to/from a string `Value` object (provided by the global template
+// functions `ConvertToValue()` and `ConvertFromValue()` defined below).
+//
+// Example usage:
+//   enum class Tp { kA, kB };
+//   template <> EnumValueDescriptor<Tp>::Description
+//   EnumValueDescriptor<Tp>::GetDescription() {
+//     return Describe("Tp").WithItem(Tp::kA, "a").WithItem(Tp::kB, "b");
+//   }
+//   ...
+//   ConvertToValue(Tp::kA, ...);
+//   ConvertFromValue(Value("b"), ...);
+template <typename T>
+class EnumValueDescriptor {
+ public:
+  static_assert(std::is_enum<T>::value,
+                "The EnumValueDescriptor parameter must be an enum");
+
+  // Class that should be used for describing items of the enum type. Should be
+  // instantiated via the `Describe()` method.
+  class Description final {
+   public:
+    Description(const char* type_name,
+                internal::EnumToValueConverter* to_value_converter,
+                internal::EnumFromValueConverter* from_value_converter)
+        : type_name_(type_name),
+          to_value_converter_(to_value_converter),
+          from_value_converter_(from_value_converter) {}
+    Description(Description&&) = default;
+    Description& operator=(Description&&) = default;
+    ~Description() = default;
+
+    const char* type_name() const { return type_name_; }
+
+    // Adds the given item into the enum's description: |enum_item_name| is the
+    // `Value` representation of |enum_item|.
+    // Returns a rvalue reference to |this| and uses the "&&" ref-qualifier, so
```

Comment:
typo: Returns ~a~ **an** rvalue

Reply:
I think a is correct here

Reply:
as a quick reference, please take a look at https://screenshot.googleplex.com/9QFaerZLyEM72G2 [Sorry, Google internal only]

Also take a look here:
https://ontariotraining.net/grammar-tip-a-or-an-with-acronyms/#:~:text=When%20it%20comes%20to%20abbreviations,first%20sound%20is%20a%20consonant.&text=Therefore%2C%20both%20%E2%80%9Ca%20Law%20Society,an%20LSUC%20form%E2%80%9D%20are%20correct

Reply:
:thumbsup: